MortgageRateVolatilityAlgorithm.cs

```diff
@@ -0,0 +1,164 @@
+                _mortgageRateStd = algorithm.STD(_mortgageRate, _indicatorPeriod, resolution);
+                _mortgageRateSma = algorithm.SMA(_mortgageRate, _indicatorPeriod, resolution);
+
+                // Use a history call to warm-up the indicators
+                WarmUpIndicators(algorithm);
+            }
+
+            public override IEnumerable<Insight> Update(QCAlgorithm algorithm, Slice data)
+            {
+                var insights = new List<Insight>();
+
+                // Return empty list if data slice doesn't contain monrtgage rate data
+                if (!data.Keys.Contains(_mortgageRate))
+                {
+                    return insights;
+                }
+                // Extract current mortgage rate, the current STD indicator value, and current SMA value
+                var rate = data[_mortgageRate].Value;
+                var deviation = _deviations * _mortgageRateStd;
+                var sma = _mortgageRateSma;
+
+                // Loop through all Active Securities to emit insights
+                foreach (var security in algorithm.ActiveSecurities.Keys)
+                {
+                    // Mortgage rate Symbol will be in the collection, so skip it
+                    if (security == _mortgageRate)
+                    {
+                        return insights;
+                    }
+
+                    // If volatility in mortgage rates is high, then we emit an Insight to sell
+                    if ((rate < sma - deviation) || (rate > sma + deviation))
+                    {
+                        insights.Add(Insight.Price(security, _insightDuration, InsightDirection.Down, _insightMagnitude));
+                    }
+
+                    // If volatility in mortgage rates is low, then we emit an Insight to buy
+                    if ((rate < sma - (decimal)deviation / 2) || (rate > sma + (decimal)deviation / 2))
+                    {
+                        insights.Add(Insight.Price(security, _insightDuration, InsightDirection.Up, _insightMagnitude));
+                    }
+                }
+
+                return insights;
+            }
+
+            private void WarmUpIndicators(QCAlgorithm algorithm)
+            {
+                // Make a history call and update the indicators
+                algorithm.History(new[] { _mortgageRate }, _indicatorPeriod, _resolution).PushThrough(bar =>
+                {
+                    _mortgageRateSma.Update(bar.EndTime, bar.Value);
+                    _mortgageRateStd.Update(bar.EndTime, bar.Value);
+                });
+            }
+        }
+        public class NasdaqMortgagePriceColumns : NasdaqDataLink
+        {
+            public NasdaqMortgagePriceColumns()
+
+                // Rename the NasdaqDataLink object column to the data we want, which is the 'Value' column
+                // of the CSV that our API call returns
+                : base(valueColumnName: "Value")
```

Comment:
I wonder if all there classes are actually now required, given we added a default `_keywords` collection?

Reply:
The problem is that some of the assets have more than one column present in `_keywords` like `SHFE/SCF2021`